lib/smooth/include/smooth/application/io/i2c/DS3231.h

```diff
@@ -0,0 +1,351 @@
+// Smooth - C++ framework for writing applications based on Espressif's ESP-IDF.
+
+// Copyright (C) 2017 Per Malmberg (https://github.com/PerMalmberg)
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+#pragma once
+
+#include <chrono>
+#include <smooth/core/Task.h>
+#include <smooth/core/io/InterruptInput.h>
+#include <smooth/core/io/i2c/I2CMasterDevice.h>
+#include <smooth/core/util/FixedBuffer.h>
+#include <memory>
+#include <utility>
+
+namespace smooth::application::io
+{
+    using namespace std::chrono;
+
+    class DS3231
+        : public core::io::i2c::I2CMasterDevice
+    {
+        private:
+            uint8_t BCD2Uint8_t(uint8_t value);
+
+            uint8_t Uint8_t2BCD(uint8_t value);
+
+        public:
+            DS3231(i2c_port_t port, uint8_t address, std::mutex& guard);
+
+            enum Register
+            {
+                DateTime = 0x00,
+                Alarm1 = 0x07,
+                Alarm2 = 0x0b,
+                Control = 0x0e,
+                Status = 0x0f,
+                AgingOffset = 0x10,
+                Temperature = 0x11
+            };
+
+            enum SquareWaveOutputFrequency
+            {
+                Frequency_1Hz = 0x00,
+                Frequency_1kHz = 0x08,
+                Frequency_4kHz = 0x10,
+                Frequency_8kHz = 0x18
+            };
+
+            enum AlarmOneRate
+            {
+                OncePerSecond = 0,
+                SecondsMatch,
+                MinutesAndSecondsMatch,
+                HoursMinutesAndSecondsMatch,
+                DateHoursMinutesAndSecondsMatch,
+                DayHoursMinutesAndSecondsMatch
+            };
+
+            enum AlarmTwoRate
+            {
+                OncePerMinute = 0,
+                MinutesMatch,
+                HoursAndMinutesMatch,
+                DateHoursAndMinutesMatch,
+                DayHoursAndMinutesMatch
+            };
+
+            //void event(const smooth::core::io::InterruptInputEvent& value) override;
```

Comment:
This is a point that I wanted to raise. I tried to get this device to accept a task so that I could fire an event if AlarmOne was triggered and another event if AlarmTwo was triggered. I got to a point where I needed a task to be passed through when the I2C device is created but could not find a solution due to my limited C++ skills.

Do you have a sample for this?
I think in the general this option would be the way the device would mostly be used but if the device is used to wake the ESP32 from deep sleep then there may need to be an option to turn off the handling of the interrupt and related events. Otherwise it may just be better to continue as is and have the user code either of the functionalities. What do you think?

Reply:
I'm guessing you had a look at the [hw_interrupt_queue](https://github.com/PerMalmberg/Smooth/tree/master/test/hw_interrupt_queue) example?

The main App is a task, which is passed as the [second argument](https://github.com/PerMalmberg/Smooth/blob/master/test/hw_interrupt_queue/hw_interrupt_queue.cpp#L27) at this line. Likewise you could pass any core::Task to InterruptQueue. The reason a Task is passed is to tell in what context the event shall be run, i.e. what task actually calls  `event(const smooth::core::io::InterruptInputEvent& value)`.

As for turning things on and off, the `core::io::InterruptInput` doesn't currently support destruction, i.e. once created it is expected to remain alive until restart of the device. That said, I do believe it would be as simple as adding a destructor and calling `gpio_isr_handler_remove(io);` Once that is in place, you could probably place an `InterruptInput` in a `std::unique_ptr` and create/destory it on the fly.

```c++
InterruptInput::~InterruptInput()
{
    gpio_isr_handler_remove(io);
}
```

Since class members are guaranteed to be valid while the destructor is running, the entire object is intact, even if the interrupt happens while the destructor is running. This doesn't hold true for derived classes so I think it is a good idea that `IntteruptInput` is made a [final](https://en.cppreference.com/w/cpp/language/final) class to prevent its use in inheritance.